File: solutions/typescript/arrays-and-hashing/longest-consecutive-sequence.ts
/**
 * 128. Longest Consecutive Sequence — https://leetcode.com/problems/longest-consecutive-sequence/
 *
 * The length of the longest run of consecutive integers present in `nums`. Position in the
 * array is irrelevant, and duplicates do not lengthen a run.
 */

import { defineProblem } from "../define-problem";

export const SLUG = "longest-consecutive-sequence";

/**
 * Sort, then walk. Time O(n log n), space O(n).
 *
 * `[...new Set(nums)].sort((a, b) => a - b)` does three jobs at once: copy (so the caller's
 * array is not reordered), deduplicate, and sort numerically. The explicit comparator is not
 * optional — the default sort compares elements as strings.
 */
export function longestConsecutiveSequenceSorting(nums: readonly number[]): number {
  if (nums.length === 0) return 0;

  const ordered = [...new Set(nums)].sort((a, b) => a - b);
  let longest = 1;
  let current = 1;

  for (let i = 1; i < ordered.length; i++) {
    current = ordered[i]! - ordered[i - 1]! === 1 ? current + 1 : 1;
    longest = Math.max(longest, current);
  }

  return longest;
}

/**
 * Walk each run exactly once, from its start. Time O(n), space O(n).
 *
 * The `!seen.has(value - 1)` guard is the entire algorithm. Without it the inner loop re-walks
 * every run from every one of its members and the whole thing degrades to O(n^2) — the version
 * people write first and then wonder why it is not faster than sorting.
 */
export function longestConsecutiveSequenceHashSet(nums: readonly number[]): number {
  const seen = new Set(nums);
  let longest = 0;

  for (const value of seen) {
    if (seen.has(value - 1)) continue; // not the start of a run

    let length = 1;
    while (seen.has(value + length)) length++;
    longest = Math.max(longest, length);
  }

  return longest;
}

export const longestConsecutiveSequence = defineProblem(SLUG, {
  sorting: longestConsecutiveSequenceSorting,
  "hash-set": longestConsecutiveSequenceHashSet,
});

File: solutions/python/arrays_and_hashing/longest_consecutive_sequence.py
"""128. Longest Consecutive Sequence — https://leetcode.com/problems/longest-consecutive-sequence/

The length of the longest run of consecutive integers present in `nums`. Position in the array
is irrelevant, and duplicates do not lengthen a run.
"""

from __future__ import annotations

from itertools import pairwise

from neetcode_core.registry import solution

SLUG = "longest-consecutive-sequence"


@solution(SLUG, approach="sorting")
def longest_consecutive_sorting(nums: list[int]) -> int:
    """Sort, then walk. Time O(n log n), space O(n).

    The trap is duplicates: `[1, 2, 2, 3]` is a run of three, not four. Deduplicating through
    a `set` before sorting removes the problem entirely, which is cheaper to reason about than
    teaching the walk to skip zero-differences.

    `itertools.pairwise` (3.10+) says "every adjacent pair" directly. The TypeScript and PHP
    versions both index manually with `ordered[i] - ordered[i - 1]`.
    """
    if not nums:
        return 0

    ordered = sorted(set(nums))
    longest = current = 1

    for previous, value in pairwise(ordered):
        current = current + 1 if value - previous == 1 else 1
        longest = max(longest, current)

    return longest


@solution(SLUG, approach="hash-set")
def longest_consecutive_hash_set(nums: list[int]) -> int:
    """Walk each run exactly once, from its start. Time O(n), space O(n).

    The `value - 1 not in seen` guard is the entire algorithm. Without it the inner loop
    re-walks every run from every one of its members and the whole thing degrades to O(n^2) —
    which is the version people write first and then wonder why it is not faster than sorting.

    With the guard, each run is walked from its smallest member and from nowhere else, so
    across the whole input the inner loop takes O(n) steps in total.
    """
    seen = set(nums)
    longest = 0

    for value in seen:
        if value - 1 in seen:
            continue  # not the start of a run — it will be counted from further down

        length = 1
        while value + length in seen:
            length += 1
        longest = max(longest, length)

    return longest
